timelock/program/src/instruction.rs

```diff
@@ -0,0 +1,590 @@
+    ///   3. `[writable]` Initialized No Voting account from which to remove your voting tokens.
+    ///   4. `[writable]` User token account that you wish your actual tokens to be returned to.
+    ///   5. `[writable]` Source holding account owned by the timelock that will has the actual tokens in escrow.
+    ///   6. `[writable]` Initialized Yes Voting dump account owned by timelock set to which to send your voting tokens.
+    ///   7. `[writable]` Initialized No Voting dump account owned by timelock set to which to send your voting tokens.
+    ///   8. `[writable]` Voting mint account.
+    ///   9. `[writable]` Yes Voting mint account.
+    ///   10. `[writable]` No Voting mint account.
+    ///   11. `[]` Timelock state account.
+    ///   12. `[]` Timelock set account.
+    ///   13. `[]` Transfer authority
+    ///   14. `[]` Yes Transfer authority
+    ///   15. `[]` No Transfer authority
+    ///   16. `[]` Timelock program mint authority
+    ///   17. `[]` Timelock program account pub key.
+    ///   18. `[]` Token program account.
+    WithdrawVotingTokens {
+        /// How many voting tokens to withdrawal
+        voting_token_amount: u64,
+    },
+
+    ///   0. `[writable]` Timelock config key. Needs to be set with pubkey set to PDA with seeds of the
+    ///           program account key, governance mint key, council mint key, timelock program account key.
+    ///   1. `[]` Program account that this config uses
+    ///   2. `[]` Governance mint that this config uses
+    ///   3. `[]` Council mint that this config uses [Optional] [Pass in 0s otherwise]
+    ///   4. `[]` Timelock program account pub key.
+    ///   5. `[]` Token program account.
+    InitTimelockConfig {
+        /// Consensus Algorithm
+        consensus_algorithm: u8,
+        /// Execution type
+        execution_type: u8,
+        /// Timelock Type
+        timelock_type: u8,
+        /// Voting entry rule
+        voting_entry_rule: u8,
+        /// Minimum slot time-distance from creation of proposal for an instruction to be placed
+        minimum_slot_waiting_period: u64,
+        /// Time limit in slots for proposal to be open to voting
+        time_limit: u64,
+        /// Optional name
+        name: [u8; CONFIG_NAME_LENGTH],
+    },
+
+    ///   0. `[]` Timelock config key. Needs to be set with pubkey set to PDA with seeds of the
+    ///           program account key, governance mint key, council mint key, and timelock program account key.
+    ///   1. `[]` Program account to tie this config to.
+    ///   2. `[]` Governance mint to tie this config to
+    ///   3. `[]` Council mint [optional] to tie this config to [Pass in 0s otherwise]
+    ///   4. `[]` Payer
+    ///   5. `[]` Timelock program account pub key.
+    ///   6. `[]` Timelock program pub key. Different from program account - is the actual id of the executable.
+    ///   7. `[]` Token program account.
+    ///   8. `[]` System account.
+    CreateEmptyTimelockConfig,
```

Comment:
Why does the timelock config need to be a PDA?

Reply:
Yes, it ensures uniqueness - nobody else can claim config for that program once this is made

Reply:
Is the worry that someone creates a competing timelock with the same program / governance / council and all that?  Is it possible for someone to grief this and give themselves some permissions and make it impossible for someone to use the combination?

Reply:
Well, we ARE going to have to preseed all the major tokens with their own configs and then hold the authorities as the foundation and grant them out as those dev teams come join governance. After that, no, it's not possible to grief them. Because the PDA includes the mint id, as long as the dev team in question creates their governance roughly at the same time as they go to create their governance token mint it's very unlikely someone is going to grief them out of it. My guess is going forward, once Governance is popularized, dev teams will actually claim their smart contract grinded key in advance, make their token mint, and claim their config well before they push any real code. It's a bit of a social change, but it will solve this quite easily.

Reply:
and yes, that IS the worry. This means that for any mint-program combo, there CAN ONLY BE ONE HIGHLANDER. :)

Reply:
Hm, it seems strange for us to pre-create these for projects and then grant them as they decide to come onboard.  Sounds like... *centralization*

Just to check, with this model, it's still possible for someone to forget to create their governance, have someone else create it and take admin control, correct?

Also, what are the gains of using the PDA over a normal new address other than uniqueness?

Reply:
Correct. The gains of the PDA are exactly that - uniqueness, and only the program itself can create it which means it can control some validation around it and own it.

I'm really not personally bothered by a bit of centralization to get a good product off the ground. After all, do you not use a JSON token list in your repo some where with mints listed for common tokens? That sounds centralized to me. Should we remove that, in the name of decentralization?

Reply:
It'll be a good product regardless of the address derivation, I'm not worried about that :smile: it's the potential for griefing that worries me, which doesn't have a solution at the moment.

The token list is an issue currently, and we do need a decentralized way of managing it.  Perhaps that's our usecase for a new `TimelockType`!

Reply:
I also had concerns about this when I saw it. It feels wrong that anybody can create a governance over any program. It is almost as if it could make the whole governance program useless because if it's successful then I can easily imagine somebody creating a front running bot which would claim governance for any new program created.
Would it be possible to require the program owner to sign the transaction to create governance over it?

Reply:
The program update authority now must sign the transaction to create governance over the program